test: complex return types in Vert.x server and Java client tests

vcrfxia opened the pull request:
### Description 

This PR improves test coverage for the Java client by exercising methods on the `Row` interface on actual rows returned from the server API. This PR also adds complex types to the tests for the server, as part of testing complex types for the client.

### Testing done 

Tests pass.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

purplefox commented on ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/ClientTest.java:
```diff
@@ -350,29 +337,66 @@ private void verifyPullQueryServerState() {
+    assertThat(row.getLong("f_long"), is(Long.valueOf(index * index)));
+    assertThat(row.getDouble("f_double"), is(index + 0.1111));
+    assertThat(row.getDecimal("f_decimal"), is(BigDecimal.valueOf(index + 0.1)));
+    assertThat(row.getKsqlArray("f_array"), is(new KsqlArray().add("s" + index).add("t" + index)));
```
I think we should also test the API of KsqlObject and KsqlArray here (or in another test case). I.e. check all the methods on those have been exercised.

vcrfxia replied:
I'm not a fan of adding comprehensive testing to KsqlObject and KsqlArray here since it feels unnecessary (a unit test should suffice) but I've gone ahead and added it anyway.

Incidentally, adding more assertions here revealed a bug in PollableSubscriber (completely unrelated to KsqlObject and KsqlArray) so I'm grateful for having done the exercise at your behest ;)

purplefox replied:
Imho, as a general rule, anything that is accessible via the API should be tested via the API. Testing only using isolated fine grained unit tests can often not expose issues that only manifest when using that part of the API like a user would. To me those are the really important tests to have.

In terms of terminology - I wouldn't say ClientTest is an integration test. Unfortunately the terminology around testing is quite poor, with "unit test" often used to mean small, fine grained tests, and integration tests where the unit is the entire system. But in reality the "unit" size is a spectrum and can vary from between these two extremes. Often the most useful kinds of tests (e.g. this ClientTest) are where the unit size is in the middle. Here the unit size is the client as a whole.

I do still think we need an integration test. By integration test I don't mean ClientTest - I consider that a coarse grained unit test. But a test that tests the client with the real(ish) server (like RestApiTest) and tests sending some real queries to it :)

vcrfxia replied:
Yeah, I've been unsure whether to call ClientTest a unit test or an integration test. If only there were another term in between, haha.

Rest assured when I've been saying "integration tests coming in a future PR" I've been using that to mean your understanding of an integration test -- testing the client with TestKsqlRestApp :)

purplefox replied:
Maybe coarse-grained unit test, or module test?

vcrfxia commented:
Added tests for KsqlObject and KsqlArray methods as suggested (and fixed a bug in PollableSubscriber while I was at it). Integration tests are coming in a follow-up PR. Thanks for the review!

vcrfxia commented:
Jenkins build is failing because of an error handling issue in one of the tests that happens to be fixed by the error handling overall in https://github.com/confluentinc/ksql/pull/5360. Since https://github.com/confluentinc/ksql/pull/5360 is stacked on top of this PR anyway, I'm going to close this one and simply merge https://github.com/confluentinc/ksql/pull/5360 instead.
